packages/navigator/src/browser/navigator-tab-bar-decorator.ts

```diff
@@ -0,0 +1,64 @@
+/********************************************************************************
+ * Copyright (C) 2020 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import { Event, Emitter } from '@theia/core/lib/common/event';
+import { TabBarDecorator } from '@theia/core/lib/browser/shell/tab-bar-decorator';
+import { EXPLORER_VIEW_CONTAINER_ID } from './navigator-widget';
+import { Title, Widget, FrontendApplicationContribution, FrontendApplication, Navigatable, Saveable, ApplicationShell } from '@theia/core/lib/browser';
+import { WidgetDecoration } from '@theia/core/lib/browser/widget-decoration';
+import { DisposableCollection, Disposable } from '@theia/core/lib/common/disposable';
+
+@injectable()
+export class NavigatorTabBarDecorator implements TabBarDecorator, FrontendApplicationContribution {
+    readonly id = 'theia-navigator-tabbar-decorator';
+    protected shell: ApplicationShell;
+
+    protected readonly emitter = new Emitter<void>();
+    private readonly toDispose = new DisposableCollection();
+    private readonly toDisposeOnDirtyChanged = new Map<Widget, Disposable>();
+
+    onStart(app: FrontendApplication): void {
```

Comment:
`onStart` does not ensure that there was no yet widgets edit

I don't think you need to implement `FrontendApplicationContribution`, instead you need to iterate over all shell widgets in `postConstruct` method and then listen to change events.

Reply:
Thank you @akosyakov for your comments. I wanted to use `shell` as an injection and use it to iterate over all the widgets in the `postConstruct` but since injecting it causes a circular dependency issue, I implemented `FrontendApplicationContribution` so that we can have access to the property of `shell` from the `FrontendApplication`, itself.

Reply:
ok, but you should still iterate over existing widgets in `onStart`, otherwise you can miss some dirty events